### PR TITLE
bf8 conv & fc layer tests added

### PR DESCRIPTION
**Pull request by nkmellem**

### CODE REVIEW

**Comment by alheinecke** [comment]
is this ready for review?

**Comment by nkmellem** [comment]
Yes, it is ready.

File: tests/conv/layer_example.c [hunk 1]
```diff
@@ -522,8 +531,12 @@ int main(int argc, char* argv[])
     /* copy out data */
     if (prec == 2) {
       libxsmm_convert_bf16_f32( output_libxsmm_bf16, output_libxsmm, nImg*nOfm*ofhp*ofwp );
+      //libxsmm_rne_convert_fp32_bf16( naive_output, naive_output_bf16, nImg*nOfm*ofhp*ofwp );
```
Comment (alheinecke) [inline tests/conv/layer_example.c]: Why not removing this?
if need to stay, please use C and not C++ comments.

File: tests/conv/layer_example.c [hunk 2]
```diff
@@ -28,10 +28,11 @@ int main(int argc, char* argv[])
   float *input_libxsmm, *filter_libxsmm, *output_libxsmm, *dinput_libxsmm, *dfilter_libxsmm, *doutput_libxsmm, *filtertr_libxsmm;
   float *bias_libxsmm;
 
-  libxsmm_bfloat16 *input_libxsmm_bf16, *filter_libxsmm_bf16, *output_libxsmm_bf16, *dinput_libxsmm_bf16, *dfilter_libxsmm_bf16, *doutput_libxsmm_bf16, *filtertr_libxsmm_bf16;
-  libxsmm_bfloat16 *bias_libxsmm_bf16;
+  libxsmm_bfloat16 *input_libxsmm_bf16, *filter_libxsmm_bf16, *output_libxsmm_bf16, *dinput_libxsmm_bf16, *dfilter_libxsmm_bf16, *doutput_libxsmm_bf16, *filtertr_libxsmm_bf16 , *bias_libxsmm_bf16;
```
Comment (alheinecke) [inline tests/conv/layer_example.c:31]: Why are changes to BF16 needed?
In fact it seems these buffers are not used?

File: tests/conv/layer_example.c [hunk 3]
```diff
@@ -589,8 +602,12 @@ int main(int argc, char* argv[])
     /* copy out data */
     if (prec == 2) {
       libxsmm_convert_bf16_f32( dinput_libxsmm_bf16, dinput_libxsmm, nImg*nIfm*ifhp*ifwp );
+      //libxsmm_rne_convert_fp32_bf16( naive_input, naive_input_bf16, nImg*nIfm*ifhp*ifwp );
```
Comment (alheinecke) [inline tests/conv/layer_example.c]: The commented code is working completely different buffers... perhaps it needs clean-up?

File: tests/conv/layer_example.c [hunk 4]
```diff
@@ -646,8 +663,12 @@ int main(int argc, char* argv[])
     }
     if (prec == 2) {
       tensor_copy_KCRSck_vnni2_to_norm_f32( dfilter_libxsmm_bf16, dfilter_libxsmm, nOfm, nIfm, kh, kw, libxsmm_dnn_conv_cfg.ifmblock, libxsmm_dnn_conv_cfg.ofmblock );
+      //libxsmm_rne_convert_fp32_bf16( naive_filter_wu, naive_filter_wu_bf16, nOfm*nIfm*kh*kw );
```
Comment (alheinecke) [inline tests/conv/layer_example.c]: dito

File: tests/conv/layer_example.c [hunk 5]
```diff
@@ -589,8 +602,12 @@ int main(int argc, char* argv[])
     /* copy out data */
     if (prec == 2) {
       libxsmm_convert_bf16_f32( dinput_libxsmm_bf16, dinput_libxsmm, nImg*nIfm*ifhp*ifwp );
+      //libxsmm_rne_convert_fp32_bf16( naive_input, naive_input_bf16, nImg*nIfm*ifhp*ifwp );
+      //libxsmm_convert_bf16_f32( naive_input_bf16, naive_input, nImg*nIfm*ifhp*ifwp );
     } else if (prec == 1) {
       libxsmm_convert_bf8_f32( dinput_libxsmm_bf8, dinput_libxsmm, nImg*nIfm*ifhp*ifwp );
+      //libxsmm_rne_convert_fp32_bf8( naive_input, naive_input_bf8, nImg*nIfm*ifhp*ifwp );
```
Comment (alheinecke) [inline tests/conv/layer_example.c]: dito

File: tests/conv/layer_example.c [hunk 6]
```diff
@@ -646,8 +663,12 @@ int main(int argc, char* argv[])
     }
     if (prec == 2) {
       tensor_copy_KCRSck_vnni2_to_norm_f32( dfilter_libxsmm_bf16, dfilter_libxsmm, nOfm, nIfm, kh, kw, libxsmm_dnn_conv_cfg.ifmblock, libxsmm_dnn_conv_cfg.ofmblock );
+      //libxsmm_rne_convert_fp32_bf16( naive_filter_wu, naive_filter_wu_bf16, nOfm*nIfm*kh*kw );
+      //libxsmm_convert_bf16_f32( naive_filter_wu_bf16, naive_filter_wu, nOfm*nIfm*kh*kw );
     } else if (prec == 1) {
       tensor_copy_KCRSck_vnni4_to_norm_f32( dfilter_libxsmm_bf8, dfilter_libxsmm, nOfm, nIfm, kh, kw, libxsmm_dnn_conv_cfg.ifmblock, libxsmm_dnn_conv_cfg.ofmblock );
+      //libxsmm_rne_convert_fp32_bf8( naive_filter_wu, naive_filter_wu_bf8, nOfm*nIfm*kh*kw );
```
Comment (alheinecke) [inline tests/conv/layer_example.c]: dito

File: tests/conv/layer_example.c [hunk 7]
```diff
@@ -28,10 +28,11 @@ int main(int argc, char* argv[])
   float *input_libxsmm, *filter_libxsmm, *output_libxsmm, *dinput_libxsmm, *dfilter_libxsmm, *doutput_libxsmm, *filtertr_libxsmm;
   float *bias_libxsmm;
 
-  libxsmm_bfloat16 *input_libxsmm_bf16, *filter_libxsmm_bf16, *output_libxsmm_bf16, *dinput_libxsmm_bf16, *dfilter_libxsmm_bf16, *doutput_libxsmm_bf16, *filtertr_libxsmm_bf16;
-  libxsmm_bfloat16 *bias_libxsmm_bf16;
+  libxsmm_bfloat16 *input_libxsmm_bf16, *filter_libxsmm_bf16, *output_libxsmm_bf16, *dinput_libxsmm_bf16, *dfilter_libxsmm_bf16, *doutput_libxsmm_bf16, *filtertr_libxsmm_bf16 , *bias_libxsmm_bf16;
+  libxsmm_bfloat16 *naive_output_bf16, *naive_filter_wu_bf16, *naive_input_bf16;
 
   libxsmm_bfloat8 *input_libxsmm_bf8, *filter_libxsmm_bf8, *output_libxsmm_bf8, *dinput_libxsmm_bf8, *dfilter_libxsmm_bf8, *doutput_libxsmm_bf8, *filtertr_libxsmm_bf8, *bias_libxsmm_bf8;
+  libxsmm_bfloat8 *naive_output_bf8, *naive_filter_wu_bf8, *naive_input_bf8;
```
Comment (alheinecke) [inline tests/conv/layer_example.c:35]: Same here, not sure why these buffers are there.

File: include/libxsmm_dnn_conv.h [hunk 8]
```diff
@@ -3,7 +3,7 @@
 * This file is part of the LIBXSMM library.                                   *
 *                                                                             *
 * For information on the license, see the LICENSE file.                       *
-* Further information: https://github.com/libxsmm/libxsmm_dnn/                *
+* Further information: https://github.com/hfp/libxsmm/                        *
```
Comment (alheinecke) [inline include/libxsmm_dnn_conv.h]: please revert you change

File: include/dnn_common.h [hunk 9]
```diff
@@ -3,7 +3,7 @@
 * This file is part of the LIBXSMM library.                                   *
 *                                                                             *
 * For information on the license, see the LICENSE file.                       *
-* Further information: https://github.com/libxsmm/libxsmm_dnn/                *
+* Further information: https://github.com/libxsmm/libxsmm/                    *
```
Comment (alheinecke) [inline include/dnn_common.h]: please revert

File: include/dnn_common.h [hunk 10]
```diff
@@ -1104,6 +1178,154 @@ LIBXSMM_INLINE void matrix_copy_KCCK_to_CKKC_bf16(libxsmm_bfloat16 *src, libxsmm
   }
 }
 
+/****************************************************************/
```
Comment (alheinecke) [inline include/dnn_common.h]: Why this comment?

File: include/dnn_common.h [hunk 11]
```diff
@@ -1104,6 +1178,154 @@ LIBXSMM_INLINE void matrix_copy_KCCK_to_CKKC_bf16(libxsmm_bfloat16 *src, libxsmm
...
+        for (k2 = 0; k2 < bk; k2++) {
+          LIBXSMM_VLA_ACCESS(5, real_dst, c1, k1, k2/4, c2, k2%4, kBlocks, bk/4, bc, 4) =
+          LIBXSMM_VLA_ACCESS(5, real_src, k1, c1, c2/4, k2, c2%4, cBlocks, bc/4, bk, 4);
+        }
+      }
+    }
+  }
+}
+
+/***************************************************************/
```
Comment (alheinecke) [inline include/dnn_common.h]: Why this comment

File: include/dnn_common.h [hunk 12]
```diff
@@ -1287,6 +1509,89 @@ LIBXSMM_INLINE void tensor_copy_NCHWc_to_NCHW_bf16(libxsmm_bfloat16 *src, libxsm
   }
 }
 
+/**********************************************************************/
```
Comment (alheinecke) [inline include/dnn_common.h]: Why this comment?

File: include/dnn_common.h [hunk 13]
```diff
@@ -1287,6 +1509,89 @@ LIBXSMM_INLINE void tensor_copy_NCHWc_to_NCHW_bf16(libxsmm_bfloat16 *src, libxsm
...
+            LIBXSMM_VLA_ACCESS(4,out,  n, c1*bc+c2, h, w, C, H, W) =
+            LIBXSMM_VLA_ACCESS(5, in, n, c1,       h, w, c2, cBlocks, H, W, bc);
+          }
+        }
+      }
+    }
+  }
+}
+
+/**********************************************************************/
```
Comment (alheinecke) [inline include/dnn_common.h]: Why this comment?

File: include/dnn_common.h [hunk 14]
```diff
@@ -1343,6 +1648,34 @@ LIBXSMM_INLINE void tensor_copy_KCRS_to_KCRSck_bf16(float *src, libxsmm_bfloat16
   }
 }
 
+LIBXSMM_INLINE void tensor_copy_KCRS_to_KCRSck_bf8(float *src, libxsmm_bfloat8 *dst, int K, int C, int R, int S, int bc, int bk)
```
Comment (alheinecke) [inline include/dnn_common.h]: it would be good to add convert to the name

File: include/dnn_common.h [hunk 15]
```diff
@@ -1371,7 +1704,7 @@ LIBXSMM_INLINE void tensor_copy_KCRSck_to_KCRS(float *src, float *dst, int K, in
   }
 }
 
-LIBXSMM_INLINE void tensor_copy_KCRSck_vnni_to_norm_f32(libxsmm_bfloat16 *src, float *dst, int K, int C, int R, int S, int bc, int bk)
+LIBXSMM_INLINE void tensor_copy_KCRSck_vnni2_to_norm_f32(libxsmm_bfloat16 *src, float *dst, int K, int C, int R, int S, int bc, int bk)
```
Comment (alheinecke) [inline include/dnn_common.h]: would be good to add convert to the name

File: include/dnn_common.h [hunk 16]
```diff
@@ -1400,6 +1733,36 @@ LIBXSMM_INLINE void tensor_copy_KCRSck_vnni_to_norm_f32(libxsmm_bfloat16 *src, f
   }
 }
 
+LIBXSMM_INLINE void tensor_copy_KCRSck_vnni4_to_norm_f32(libxsmm_bfloat8 *src, float *dst, int K, int C, int R, int S, int bc, int bk)
```
Comment (alheinecke) [inline include/dnn_common.h]: would be good to add convert to the name

File: include/dnn_common.h [hunk 17]
```diff
@@ -1456,6 +1819,34 @@ LIBXSMM_INLINE void tensor_transpose_KCRSck_to_CKRSkc_bf16(float *src, libxsmm_b
   }
 }
 
+LIBXSMM_INLINE void tensor_transpose_KCRSck_to_CKRSkc_bf8(float *src, libxsmm_bfloat8 *dst, int K, int C, int R, int S, int bc, int bk)
```
Comment (alheinecke) [inline include/dnn_common.h]: would be good to add convert to the name

File: include/libxsmm_dnn_conv.h [hunk 18]
```diff
@@ -169,6 +169,24 @@ typedef struct libxsmm_dnn_conv_config {
   libxsmm_meltwfunction_unary  relu_kernel_bf16;
   libxsmm_meltwfunction_binary colbias_add_kernel_bf16;
 
+  libxsmm_xmmfunction fwd_compute_kernel_strd_fused_bf8;
```
Comment (alheinecke) [inline include/libxsmm_dnn_conv.h:172]: xmmfunction, meltwfunction are polymoprh, why can we use the same variables, slightly renamed for bf16 and bf8?

File: src/libxsmm_dnn_conv.c [hunk 19]
```diff
@@ -3,7 +3,7 @@
 * This file is part of the LIBXSMM library.                                   *
 *                                                                             *
 * For information on the license, see the LICENSE file.                       *
-* Further information: https://github.com/libxsmm/libxsmm_dnn/                *
+* Further information: https://github.com/hfp/libxsmm/                        *
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv.c]: please revert

File: src/libxsmm_dnn_conv.c [hunk 20]
```diff
@@ -434,7 +445,7 @@ LIBXSMM_API_INLINE int libxsmm_dnn_conv_setup_init_fwd_gemm_flags( libxsmm_dnn_c
     result = LIBXSMM_GEMM_FLAG_ALIGN_C_NTS_HINT;
   }
   /* Disable since the GEMM output is going to f32 scratch  */
-  if (cfg->datatype_in == LIBXSMM_DATATYPE_BF16 || cfg->datatype_in == LIBXSMM_DATATYPE_I8) {
+  if (cfg->datatype_in == LIBXSMM_DATATYPE_BF16 || cfg->datatype_in == LIBXSMM_DATATYPE_BF8 || cfg->datatype_in == LIBXSMM_DATATYPE_I8) {
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv.c:435]: This condition is joined with BF16, there are others which aren't.. I'm fine with combining, but then we should do everywhere.

File: src/libxsmm_dnn_conv.c [hunk 21]
```diff
@@ -289,6 +291,11 @@ LIBXSMM_API_INLINE int libxsmm_dnn_conv_setup_blocksifm_blocking( libxsmm_dnn_co
   if (cfg->datatype_in == LIBXSMM_DATATYPE_BF16) {
     result = cfg->blocksifm;
   }
+
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv.c]: see the comment on combining such ifs

File: src/libxsmm_dnn_conv.c [hunk 22]
```diff
@@ -377,6 +384,10 @@ LIBXSMM_API_INLINE int libxsmm_dnn_conv_setup_avoid_rim_fmas_fwd( libxsmm_dnn_co
     result = 0;
   }
 
+  if (cfg->datatype_in == LIBXSMM_DATATYPE_BF8) {
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv.c]: see the comment on combining such ifs

File: src/libxsmm_dnn_conv.c [hunk 23]
```diff
@@ -625,6 +636,10 @@ LIBXSMM_API_INLINE int libxsmm_dnn_conv_setup_blocksofm_blocking( libxsmm_dnn_co
     result = cfg->blocksofm;
   }
 
+  if (cfg->datatype_in == LIBXSMM_DATATYPE_BF8) {
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv.c]: see the comment on combining such ifs

File: src/libxsmm_dnn_conv.c [hunk 24]
```diff
@@ -2830,7 +3871,10 @@ LIBXSMM_API libxsmm_dnn_conv_config setup_libxsmm_dnn_conv( libxsmm_datatype cnn
   res.upd_padding_copy = libxsmm_dnn_conv_setup_upd_padding_copy(&res);
 
   if (cnn_dtype_in == LIBXSMM_DATATYPE_BF16) {
+  //if (cnn_dtype_in == LIBXSMM_DATATYPE_BF16 || cnn_dtype_in == LIBXSMM_DATATYPE_BF8) {
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv.c]: can this be removed?

File: src/libxsmm_dnn_conv_fwd_custom_custom_bf8.c [hunk 25]
```diff
@@ -0,0 +1,547 @@
+/******************************************************************************
+* Copyright (c) Intel Corporation - All rights reserved.                      *
+* This file is part of the LIBXSMM library.                                   *
+*                                                                             *
+* For information on the license, see the LICENSE file.                       *
+* Further information: https://github.com/hfp/libxsmm/                        *
```
Comment (alheinecke) [inline src/libxsmm_dnn_conv_fwd_custom_custom_bf8.c]: this should be libxsmm/libxsmm_dnn as in the other files

File: tests/conv/layer_example.c [hunk 26]
```diff
@@ -3,7 +3,7 @@
 * This file is part of the LIBXSMM library.                                   *
 *                                                                             *
 * For information on the license, see the LICENSE file.                       *
-* Further information: https://github.com/libxsmm/libxsmm_dnn/                *
+* Further information: https://github.com/hfp/libxsmm/                        *
```
Comment (alheinecke) [inline tests/conv/layer_example.c]: please revert

File: tests/conv/layer_example.c [hunk 27]
```diff
@@ -147,7 +158,7 @@ int main(int argc, char* argv[])
   if (argc > i) fuse_type = atoi(argv[i++]);
   if (argc > i) bc = atoi(argv[i++]);
   if (argc > i) bk = atoi(argv[i++]);
-  if (argc > i) prec_bf16 = atoi(argv[i++]);
+  if (argc > i) prec = atoi(argv[i++]);
```
Comment (alheinecke) [inline tests/conv/layer_example.c:161]: better to use char and then mapping the libxsmm_datatype as this solution doesn't scale to HF8

File: tests/conv/run_convs.sh [hunk 28]
```diff
@@ -101,378 +101,380 @@ fi
 # ./layer_example iters inpWidth inpHeight nImg nIfm nOfm kw kh padw padh stride type
 #
 if [ "f32" == "${BIN}" ]; then
-  PREC_BF16=0
+  PREC=4
```
Comment (alheinecke) [inline tests/conv/run_convs.sh:104]: see comment on using a string "F32, BF16, BF8" in the driver as this prepares better for HF8

File: tests/fc/run_fullyconnected.sh [hunk 29]
```diff
@@ -98,13 +98,14 @@ if [ "" = "${LIBXSMM_TARGET_HIDDEN}" ] || [ "0" = "${LIBXSMM_TARGET_HIDDEN}" ];
 fi
 
 if [ "f32" == "${BIN}" ]; then
-  PREC_BF16=0
```
Comment (alheinecke) [inline tests/fc/run_fullyconnected.sh:101]: See comment on string "F32, Bf16, BF8" in the driver as this prepares better for HF8

File: tests/fc/layer_example.c [hunk 30]
```diff
@@ -131,7 +155,7 @@ int main(int argc, char* argv[])
   if (argc > i) bn         = atoi(argv[i++]);
   if (argc > i) bk         = atoi(argv[i++]);
   if (argc > i) bc         = atoi(argv[i++]);
-  if (argc > i) prec_bf16  = atoi(argv[i++]);
+  if (argc > i) prec       = atoi(argv[i++]);
```
Comment (alheinecke) [inline tests/fc/layer_example.c:158]: better to use char and then mapping to LIBXSMM_DATETYPE as this solution won't allow for HF8

**Review by alheinecke — CHANGES_REQUESTED** [review]
Minor changes, mainly clean-up (wrong links, some commented code should be removed, precision as char + LIBXSMM_DATATYPE to be future proof with HF8)

File: tests/conv/layer_example.c (hunk 26 above)
Comment (nkmellem) [reply tests/conv/layer_example.c]: done

File: include/dnn_common.h (hunk 14 above)
Comment (nkmellem) [reply include/dnn_common.h]: done.

File: include/dnn_common.h (hunk 15 above)
Comment (nkmellem) [reply include/dnn_common.h]: done

File: include/dnn_common.h (hunk 17 above)
Comment (nkmellem) [reply include/dnn_common.h]: done

File: src/libxsmm_dnn_conv.c (hunk 19 above)
Comment (nkmellem) [reply src/libxsmm_dnn_conv.c]: done

File: src/libxsmm_dnn_conv.c (hunk 20 above)
Comment (nkmellem) [reply src/libxsmm_dnn_conv.c:435]: combined. resolved

File: src/libxsmm_dnn_conv.c (hunk 24 above)
Comment (nkmellem) [reply src/libxsmm_dnn_conv.c]: done

File: tests/conv/layer_example.c (hunk 27 above)
Comment (nkmellem) [reply tests/conv/layer_example.c:161]: strcmp is messy. postponed this change to a later date

**Comment by nkmellem** [comment]
comments resolved

**Comment by hfp** [comment]
I have enabled your account such that PRs are tested even when sent from a forked repository.

**Comment by hfp** [comment]
Can you please merge the current main version of libxsmm_dnn? Since we split-out the repo from libxsmm, I worked on getting CI up again. Now, your PR misses these changes. I wanted to see of your PR would become "green"...

**Pull request merged**